Plot.fn()

berekuk opened the pull request:
Also, accidental discovery: methods on Squiggle objects are easy to implement, won't even need a parser change.

We can have `arr.length` and `dist.inv(0.5)` and so on. (see the refactoring from `$_atIndex_$` to `get` method on Value objects).

TODO:
- [x] disable minX/maxX fields in UI settings for function plots, or make them override the defaults
- [x] fix tests
- [x] documentation
- [x] support log X scale?

OAGr commented on packages/components/src/components/SquiggleViewer/ExpressionViewer.tsx:
```diff
@@ -202,10 +202,15 @@ export const ExpressionViewer: React.FC<Props> = ({ value }) => {
+                      pointSet.ok && hasMassBelowZero(pointSet.value.asShape())
+                    );
+                  })
+                : true; // TODO - check min/max values for fn plots and support log scale in FunctionChart
```
just to be clear, this is a TODO for later?

berekuk replied:
Log scales were never implemented for function charts, so it's fine to do it later, but also not hard to do in this PR, I'll probably do it.

berekuk replied:
Wait, this setting is in distribution section, so for function charts it works for distributions that show on mouseover. So it should be true, without `TODO`.

OAGr commented on packages/squiggle-lang/src/library/index.ts:
```diff
@@ -18,38 +13,19 @@ const makeStdLib = (): Namespace => {
-            ErrorMessage.throw(
-              RERecordPropertyNotFound("Record property not found", index)
-            )
-          );
```
curious - what was going on here?

berekuk replied:
This code was responsible for field lookups: `arr[3]`, `rec.foo` and `rec["foo"]`.

I needed to access `plotObj.fn` for rendering the function, so I had to make it more polymorphic and it made sense to move the code to `value.get` accessors on each value class.

Or, if you mean `??`, it's a https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing operator.

OAGr replied:
I meant the former, thanks!

OAGr commented on packages/squiggle-lang/src/value/index.ts:
```diff
@@ -135,17 +178,46 @@ export type LabeledDistribution = {
+  | {
+      type: "fn";
+      fn: Lambda;
+      min: number;
```
Tiny point, but maybe min/max should be optional? Min=0, max=10.

berekuk replied:
I considered this when I wrote the code, but right now we don't have the support in `frRecord` for optional fields. It could be done with duplicating the `Plot.fn` definition in all combinations, but I don't think it's worth the hassle.

I agree that it needs to be implemented if `Plot.fn` gets more arguments (color, etc).

OAGr replied:
Makes sense, let's ignore for now.

OAGr commented on packages/squiggle-lang/src/value/index.ts:
```diff
@@ -135,17 +178,46 @@ export type LabeledDistribution = {
+      max: number;
+    };
+
+class VPlot implements Indexable {
```
I realize this was there before, but ``VPlot`` seems like a bad name, little clue what it means looking at this PR

berekuk replied:
Do you dislike the "Plot" name or "V" prefix? The prefix is common for all value objects, and since they're so common I don't want to make it longer; "PlotValue" would be worse, I think.

OAGr replied:
Ah, didn't realize there was a convention. That's fine then.

OAGr commented on packages/squiggle-lang/src/value/index.ts:
```diff
@@ -26,12 +37,29 @@ If you add a new value class, don't forget to add it to the "Value" union type b
+
+  get(key: Value) {
+    if (key.type === "Number") {
+      const index = key.value | 0; // TODO - fail on non-integer indices?
```
> / TODO - fail on non-integer indices?

That seems good to me

berekuk replied:
I'll fix this is a separate PR.

OAGr commented on packages/squiggle-lang/src/value/index.ts:
```diff
@@ -26,12 +37,29 @@ If you add a new value class, don't forget to add it to the "Value" union type b
+    if (key.type === "Number") {
+      const index = key.value | 0; // TODO - fail on non-integer indices?
+      if (index >= 0 && index < this.value.length) {
+        return this.value[index];
```
Might be obvious - but won't `` this.value[index]`` return an error if out of bounds, that you could use? Instead of redoing that check?

berekuk replied:
No, out of bounds index lookup on arrays returns `undefined` in JS. It would probably fail later but with a cryptic error.

OAGr commented on packages/squiggle-lang/src/value/index.ts:
```diff
@@ -135,17 +178,46 @@ export type LabeledDistribution = {
+        return `Plot for function ${this.value.fn}`;
+      default:
+        // never happens
+        return "Unknown plot";
```
Maybe throw error

berekuk replied:
Changed to explicit return type, typescript should handle the exhaustiveness check for us.

OAGr commented on packages/squiggle-lang/src/fr/plot.ts:
```diff
@@ -46,11 +48,33 @@ export const library = [
+    definitions: [
+      makeDefinition(
+        "fn",
+        [frRecord(["fn", frLambda], ["min", frNumber], ["max", frNumber])],
```
you might want to add to Documentation, I assume that's not automatically done right now.

OAGr reviewed (approved):
This seems good to me. I don't totally understand it, but nothing more bad/medicore stands out.

Looking forward to using this!

OAGr commented:
Now that I think about it, one issue is that this is called ``Plot.fn``, but it works for 1-float functions. This is probably an okay default, but I'm sure we'll soon want other options for other cases - like functions of 2 variables, for example.

berekuk commented:
I added support for rendering `Function.declare` as well:
https://squiggle-website-git-custom-plots-quantified-uncertainty.vercel.app/playground#code=eNo1yb0OwiAUBtBX%2BXInNejSuNzV6As42g6k0EhSLoZCxFDe3Z%2FE9ZxKyz08r9l7HV%2FEKWarfnQ2LoX4F2Mnned0CsYS0yXLmFyQg7HjrKPd1F6ASRgVa1lRdgVNfc3JI6eFcavw7vP7TsHrwjiiDb20LbU34kAseQ%3D%3D

Also fixed settings:
<img width="536" alt="Captura de pantalla 2023-03-24 a la(s) 17 01 18" src="https://user-images.githubusercontent.com/89368/227658747-f5ea9909-2904-4ccb-8b9c-389ba4f5b509.png">
